viz/LaserLine.cpp

```diff
@@ -2,44 +2,123 @@
 }
 
 osg::ref_ptr<osg::Node> LaserLine::createMainNode()
 {
-    // Geode is a common node used for vizkit3d plugins. It allows to display
-    // "arbitrary" geometries
-    return new osg::Geode();
+    return lineHolder->getRootNode();
 }
 
-void LaserLine::updateMainNode ( osg::Node* node )
+void LaserLine::updateMainNode (osg::Node* node)
 {
-    osg::Geode* geode = static_cast<osg::Geode*>(node);
-    // Update the main node using the data in p->data
+    lineHolder->update(vertices, color);
 }
 
 void LaserLine::updateDataIntern(base::samples::LaserScan const& value)
 {
-    p->data = value;
-    std::cout << "got new sample data" << std::endl;
+    loadLineVerticesFromLaserScan(value);
+}
+
+bool LaserLine::isShowRange() const
+{
+    return lineHolder->isShowRange();
+}
+void LaserLine::setShowRange(bool value)
+{
+    lineHolder->setShowRange(value);
+}
+
+QColor LaserLine::getColor()
+{
+    QColor qcolor;
+    qcolor.setRgbF(color[0], color[1], color[2], color[3]);
+    return qcolor;
+}
+
+void LaserLine::setColor(QColor value)
+{
+    if (value.isValid()) {
+        color.set(value.redF(),
+                  value.greenF(),
+                  value.blueF(),
+                  value.alphaF());
+    }
+}
+
+double LaserLine::getLineWidth()
+{
+    return (double)lineHolder->getLineWidth();
+}
+
+void LaserLine::setLineWidth(double lineWidth)
+{
+    lineHolder->setLineWidth((float)lineWidth);
+}
+
+bool LaserLine::getShowAllLines()
+{
+    return showAllLines;
+}
+
+void LaserLine::setShowAllLines(bool value)
+{
+    showAllLines = value;
+}
+
+bool LaserLine::loadLineVerticesFromLaserScan(base::samples::LaserScan const& laserScan)
+{
+    std::vector<uint32_t> ranges = laserScan.ranges;
+
+    double angle = laserScan.start_angle; //start angle
+    double angleStep = laserScan.angular_resolution; //angle step
+    double maxRange = laserScan.maxRange / 1000.0; //convert millimeters to meters
+
+    vertices->clear();
+
+    if (!ranges.empty() && ranges.size() >= 2) {
+
+
+        for (int i = 0; i < ranges.size()-1; i++) {
+
+            //get the max range point
+            Eigen::Vector3d limit = Eigen::Quaterniond(Eigen::AngleAxisd(angle + (i * angleStep), Eigen::Vector3d::UnitZ())) * Eigen::Vector3d(maxRange, 0, 0.0);
+
+            /**
+             * get the start and end line points
+             * compute the rotation using the angle
+             */
+            Eigen::Vector3d pt0 = Eigen::Quaterniond(Eigen::AngleAxisd(angle + (i * angleStep), Eigen::Vector3d::UnitZ())) * Eigen::Vector3d((ranges[i]) / 1000.0, 0, 0.0);
+            Eigen::Vector3d pt1 = Eigen::Quaterniond(Eigen::AngleAxisd(angle + ((i+1) * angleStep), Eigen::Vector3d::UnitZ())) * Eigen::Vector3d((ranges[i+1]) / 1000.0, 0, 0.0);
+
+            /**
+             * Filter occlusion points and points with values equal the max range
+             */
+            if ((showAllLines) || (abs(pt0[0] - pt1[0]) < 0.1 && (pt0[0] <  limit[0] && pt1[0] < limit[0]))) {
```

Comment:
This is ... weird. If you want to know if pt0 or pt1 are beyond max range, then test for it. Why are you testing only x ? What are you trying to achieve ?